### PR TITLE
add tests for Enzyme frontend for nonlinear problem

### PR DESCRIPTION
**Pull request by m-bossart**
Add tests for SciML/NonlinearSolve.jl#439

### CODE REVIEW

File: test/steady_state.jl [hunk 1]
```diff
@@ -427,14 +427,39 @@ end
     test_loss(p, prob4, alg = DynamicSS(Rodas5()))
     test_loss(p, prob2, alg = SimpleNewtonRaphson())
 
+    function enzyme_gradient(p, prob; alg = NewtonRaphson())
+        dp = Enzyme.make_zero(p)
+        dprob = Enzyme.make_zero(prob)
+        Enzyme.autodiff(Reverse, (p, prob) -> test_loss(p, prob, alg=alg), Active, Duplicated(p,dp), Duplicated(prob,dprob))
```
Comment (wsmoses) [inline test/steady_state.jl:433]: You should just pass Alg as an extra arg if possible rather than making a closure over it

**Comment by ChrisRackauckas** [comment]
Where is this at?

**Comment by m-bossart** [comment]
I made the change to not use a closure put pass all the arguments to autodiff. I included the tests for all the cases that we included previously with Zygote but not all of them pass (the commented ones). The uncommented tests with Enzyme should pass

**Comment by ChrisRackauckas** [comment]
I mean, what needs to review and merge first?

**Comment by m-bossart** [comment]
First SciML/SimpleNonlinearSolve.jl#151, then this should be ready.

File: src/adjoint_common.jl [hunk 2]
```diff
@@ -43,7 +43,7 @@ function adjointdiffcache(g::G, sensealg, discrete, sol, dgdu::DG1, dgdp::DG2, f
     else
         throw(SciMLStructuresCompatibilityError())
     end
-    if prob isa Union{SteadyStateProblem, NonlinearProblem}
+    if prob isa Union{SteadyStateProblem, ImmutableNonlinearProblem}
```
Comment (ChrisRackauckas) [inline src/adjoint_common.jl:46]:
```suggestion
    if prob isa Union{SteadyStateProblem, NonlinearProblem, ImmutableNonlinearProblem}
```

File: src/adjoint_common.jl [hunk 3]
```diff
@@ -61,7 +61,7 @@ function adjointdiffcache(g::G, sensealg, discrete, sol, dgdu::DG1, dgdp::DG2, f
         _W = nothing
     end
 
-    if prob isa Union{SteadyStateProblem, NonlinearProblem}
+    if prob isa Union{SteadyStateProblem, ImmutableNonlinearProblem}
```
Comment (ChrisRackauckas) [inline src/adjoint_common.jl:64]:
```suggestion
    if prob isa Union{SteadyStateProblem, NonlinearProblem, ImmutableNonlinearProblem}
```

File: src/adjoint_common.jl [hunk 4]
```diff
@@ -179,7 +179,7 @@ function adjointdiffcache(g::G, sensealg, discrete, sol, dgdu::DG1, dgdp::DG2, f
     @assert autojacvec !== nothing
 
     if autojacvec isa ReverseDiffVJP
-        if prob isa Union{SteadyStateProblem, NonlinearProblem}
+        if prob isa Union{SteadyStateProblem, ImmutableNonlinearProblem}
```
Comment (ChrisRackauckas) [inline src/adjoint_common.jl:182]:
```suggestion
        if prob isa Union{SteadyStateProblem, NonlinearProblem, ImmutableNonlinearProblem}
```

File: src/SciMLSensitivity.jl [hunk 5]
```diff
@@ -19,7 +19,7 @@ using SciMLOperators
 using Functors
 import TruncatedStacktraces
 import SciMLStructures
-
+import SimpleNonlinearSolve: ImmutableNonlinearProblem
```
Comment (ChrisRackauckas) [inline src/SciMLSensitivity.jl:22]:
```suggestion
```

File: Project.toml [hunk 6]
```diff
@@ -35,6 +35,7 @@ ReverseDiff = "37e2e3b7-166d-5795-8a7a-e32c996b4267"
 SciMLBase = "0bca4576-84f4-4d90-8ffe-ffa030f20462"
 SciMLOperators = "c0aeaf25-5076-4817-a8d5-81caf7dfa961"
 SciMLStructures = "53ae85a6-f571-4167-b2af-e1d143709226"
+SimpleNonlinearSolve = "727e6d20-b764-4bd8-a329-72de5adea6c7"
```
Comment (ChrisRackauckas) [inline Project.toml:38]:
```suggestion
```

File: src/steadystate_adjoint.jl [hunk 7]
```diff
@@ -37,7 +37,7 @@ end
         kwargs...) where {DG1, DG2, G}
     @unpack f, p, u0 = sol.prob
 
-    sol.prob isa NonlinearProblem && (f = ODEFunction(f))
+    sol.prob isa ImmutableNonlinearProblem && (f = ODEFunction(f))
```
Comment (ChrisRackauckas) [inline src/steadystate_adjoint.jl:40]:
```suggestion
    sol.prob isa AbstractNonlinearProblem && (f = ODEFunction(f))
```

File: src/derivative_wrappers.jl [hunk 8]
```diff
@@ -498,13 +498,13 @@ function _vecjacobian!(dλ, y, λ, p, t, S::TS, isautojacvec::ReverseDiffVJP, dg
...
     ReverseDiff.unseed!(tp)
-    if !(prob isa Union{SteadyStateProblem, NonlinearProblem})
+    if !(prob isa Union{SteadyStateProblem, ImmutableNonlinearProblem})
         ReverseDiff.unseed!(tt)
     end
     W !== nothing && ReverseDiff.unseed!(tW)
     ReverseDiff.value!(tu, y)
     p isa DiffEqBase.NullParameters || ReverseDiff.value!(tp, tunables)
-    if !(prob isa Union{SteadyStateProblem, NonlinearProblem})
+    if !(prob isa Union{SteadyStateProblem, ImmutableNonlinearProblem})
```
Comment (ChrisRackauckas) [inline src/derivative_wrappers.jl:507]:
```suggestion
    if !(prob isa AbstractNonlinearProblem)
```

File: src/derivative_wrappers.jl [hunk 9]
```diff
@@ -498,13 +498,13 @@ function _vecjacobian!(dλ, y, λ, p, t, S::TS, isautojacvec::ReverseDiffVJP, dg
     output = ReverseDiff.output_hook(tape)
     ReverseDiff.unseed!(tu) # clear any "leftover" derivatives from previous calls
     ReverseDiff.unseed!(tp)
-    if !(prob isa Union{SteadyStateProblem, NonlinearProblem})
+    if !(prob isa Union{SteadyStateProblem, ImmutableNonlinearProblem})
```
Comment (ChrisRackauckas) [inline src/derivative_wrappers.jl:501]:
```suggestion
    if !(prob isa AbstractNonlinearProblem)
```

File: src/derivative_wrappers.jl [hunk 10]
```diff
@@ -486,7 +486,7 @@ function _vecjacobian!(dλ, y, λ, p, t, S::TS, isautojacvec::ReverseDiffVJP, dg
         end
     end
 
-    if prob isa Union{SteadyStateProblem, NonlinearProblem}
+    if prob isa Union{SteadyStateProblem, ImmutableNonlinearProblem}
```
Comment (ChrisRackauckas) [inline src/derivative_wrappers.jl:489]:
```suggestion
    if prob isa AbstractNonlinearProblem
```

File: src/derivative_wrappers.jl [hunk 11]
```diff
@@ -447,7 +447,7 @@ function _vecjacobian!(dλ, y, λ, p, t, S::TS, isautojacvec::ReverseDiffVJP, dg
     end
 
     u0 = state_values(prob)
-    if prob isa Union{SteadyStateProblem, NonlinearProblem} ||
+    if prob isa Union{SteadyStateProblem, ImmutableNonlinearProblem} ||
```
Comment (ChrisRackauckas) [inline src/derivative_wrappers.jl:450]:
```suggestion
    if prob isa AbstractNonlinearProblem ||
```

File: src/derivative_wrappers.jl [hunk 12]
```diff
@@ -250,7 +250,7 @@ function _vecjacobian!(dλ, y, λ, p, t, S::TS, isautojacvec::Bool, dgrad, dy,
         J = get_tmp(J, dλ)
     end
 
-    if !(prob isa Union{SteadyStateProblem, NonlinearProblem}) && dλ !== nothing
+    if !(prob isa Union{SteadyStateProblem, ImmutableNonlinearProblem}) && dλ !== nothing
```
Comment (ChrisRackauckas) [inline src/derivative_wrappers.jl:253]:
```suggestion
    if !(prob isa AbstractNonlinearProblem) && dλ !== nothing
```

File: src/concrete_solve.jl [hunk 13]
```diff
@@ -319,7 +319,7 @@ function DiffEqBase._concrete_solve_adjoint(
 end
 
 const ADJOINT_STEADY_PROBLEM_ERROR_MESSAGE = """
-                                             Chosen adjoint method is not compatible with the chosen problem. NonlinearProblem
+                                             Chosen adjoint method is not compatible with the chosen problem. ImmutableNonlinearProblem
```
Comment (ChrisRackauckas) [inline src/concrete_solve.jl:322]:
```suggestion
                                             Chosen adjoint method is not compatible with the chosen problem. NonlinearProblem
```

File: src/adjoint_common.jl [hunk 14]
```diff
@@ -179,7 +179,7 @@ function adjointdiffcache(g::G, sensealg, discrete, sol, dgdu::DG1, dgdp::DG2, f
     @assert autojacvec !== nothing
 
     if autojacvec isa ReverseDiffVJP
-        if prob isa Union{SteadyStateProblem, NonlinearProblem}
+        if prob isa Union{SteadyStateProblem, NonlinearProblem, NonlinearLeastSquaresProblem, ImmutableNonlinearProblem}
```
Comment (ChrisRackauckas) [inline src/adjoint_common.jl:182]:
```suggestion
        if prob isa AbstractNonlinearProblem
```

File: src/adjoint_common.jl [hunk 15]
```diff
@@ -43,7 +43,7 @@ function adjointdiffcache(g::G, sensealg, discrete, sol, dgdu::DG1, dgdp::DG2, f
     else
         throw(SciMLStructuresCompatibilityError())
     end
-    if prob isa Union{SteadyStateProblem, NonlinearProblem}
+    if prob isa Union{SteadyStateProblem, NonlinearProblem, NonlinearLeastSquaresProblem, ImmutableNonlinearProblem}
```
Comment (ChrisRackauckas) [inline src/adjoint_common.jl:46]:
```suggestion
    if prob isa AbstractNonlinearProblem
```

File: src/adjoint_common.jl [hunk 16]
```diff
@@ -61,7 +61,7 @@ function adjointdiffcache(g::G, sensealg, discrete, sol, dgdu::DG1, dgdp::DG2, f
         _W = nothing
     end
 
-    if prob isa Union{SteadyStateProblem, NonlinearProblem}
+    if prob isa Union{SteadyStateProblem, NonlinearProblem, NonlinearLeastSquaresProblem, ImmutableNonlinearProblem}
```
Comment (ChrisRackauckas) [inline src/adjoint_common.jl:64]:
```suggestion
    if prob isa AbstractNonlinearProblem
```

**Pull request merged**